File: android/src/main/java/com/apollohg/editor/viewer/PreparedProseLayout.kt
package com.apollohg.editor.viewer

import android.graphics.Rect
import android.text.StaticLayout
import com.apollohg.editor.ProseViewerError

internal data class ProseLayoutKey(
    val semanticKey: String,
    val widthPx: Int,
    val themeDigest: String,
    val nativeFontRevision: Long,
    val fontEnvironmentRevision: Long,
    val densityBits: Long,
    val attachmentRevision: Long,
    val generationIdentity: String,
    /** Immutable semantic diagnostic context; excludes replacement revisions. */
    val semanticGenerationIdentity: String = semanticKey
)

internal data class FabricSurfaceToken(val surfaceId: Int, val componentTag: Int)

/**
 * The state-owned Fabric incarnation.  A recycled component may reuse the
 * same surface/tag and even the same semantic generation, so every handoff
 * operation must carry the native lease handle as well.
 */
internal data class FabricGenerationToken(
    val surface: FabricSurfaceToken,
    val generationIdentity: String,
    val leaseHandle: Long
) {
    init {
        require(leaseHandle > 0) { "Fabric lease handles must be signed, non-zero Int64 values." }
    }
}

/** State-lifetime identity shared by all semantic/width revisions of a family. */
internal data class FabricLeaseOwner(val surface: FabricSurfaceToken, val leaseHandle: Long)
internal data class FabricLeaseKey(
    val generation: FabricGenerationToken,
    val layout: ProseLayoutKey
)
internal data class ProseMountKey(
    val generationIdentity: String,
    val widthPx: Int,
    val densityBits: Long
)

/** Paint-only operations which Android spans cannot represent accurately. */
internal enum class PreparedProseFragmentKind {
    TEXT,
    MARKER,
    BACKGROUND,
    BORDER,
    RULE,
    ATOM,
    STRIKE,
    IMAGE
}

internal data class PreparedProseFragment(
    val kind: PreparedProseFragmentKind,
    val bounds: Rect,
    val layout: StaticLayout? = null,
    val layoutX: Int = bounds.left,
    val layoutY: Int = bounds.top,
    val labelLayout: StaticLayout? = null,
    val labelX: Int = bounds.left,
    val labelY: Int = bounds.top,
    val color: Int? = null,
    val borderColor: Int? = null,
    val cornerRadius: Float = 0f,
    val strokeWidth: Float = 0f,
    val label: String? = null,
    /** Immutable compiler identity for an inline atom; absent for non-atoms. */
    val atomNodeType: String? = null,
    val atomDocPos: Long? = null,
    val atomAttrsJson: String? = null,
    val checked: Boolean = false,
    val box: com.apollohg.editor.EditorBoxStyle? = null,
    val decorationBounds: Rect? = null,
    val resizeMode: String = "contain"
) {
    val retainedBytes: Long
        get() = 160L + (layout?.text?.length ?: 0).toLong() * 4 +
            (labelLayout?.text?.length ?: 0).toLong() * 4 +
            (label?.length ?: 0).toLong() * 2 +
            (atomNodeType?.length ?: 0).toLong() * 2 +
            (atomAttrsJson?.length ?: 0).toLong() * 2
}

/** An immutable culling unit in document paint order. */
internal data class PreparedProseBlock(
    val fragments: List<PreparedProseFragment>,
    val bounds: Rect
) {
    val topPx: Int get() = bounds.top
    val bottomPx: Int get() = bounds.bottom
    fun intersects(clip: Rect): Boolean = bottomPx > clip.top && topPx < clip.bottom
    val retainedBytes: Long get() = 160L + fragments.sumOf { it.retainedBytes }
}

internal data class PreparedProseInteraction(
    val kind: Kind,
    val rects: List<Rect>,
    val href: String? = null,
    val visibleText: String,
    /** Kept as Long so the complete unsigned Rust u32 domain is lossless. */
    val docPos: Long? = null,
    val label: String,
    val attrsJson: String? = null
) {
    enum class Kind { LINK, MENTION }
    val retainedBytes: Long get() = 144L + rects.size * 32L + (href?.length ?: 0) * 2L +
        visibleText.length * 2L +
        label.length * 2L +
        (attrsJson?.length ?: 0) * 2L
}

internal data class PreparedProseAccessibilityNode(
    val interactionIndex: Int,
    val role: Role,
    val label: String,
    val bounds: Rect
) {
    enum class Role { LINK, MENTION }
    val retainedBytes: Long get() = 96L + label.length * 2L
}

/** A fully prepared artifact. StaticLayout construction is complete before publication. */
internal data class PreparedProseLayout(
    val key: ProseLayoutKey,
    val widthPx: Int,
    val heightPx: Int,
    val blocks: List<PreparedProseBlock>,
    val interactions: List<PreparedProseInteraction> = emptyList(),
    val accessibilityNodes: List<PreparedProseAccessibilityNode> = emptyList(),
    val imageAttachments: List<ViewerImageAttachment> = emptyList(),
    val retainedBytes: Long,
    val error: ProseViewerError? = null,
    val viewerAtoms: List<PreparedViewerAtom> = emptyList(),
    val contentBox: com.apollohg.editor.EditorBoxStyle? = null,
    val codeHighlighting: com.apollohg.editor.NativeCodeHighlightingConfig? = null,
    val codeHighlightBlocks: List<com.apollohg.editor.CodeHighlightBlock> = emptyList(),
    val highlightedCodeKeys: Set<String> = emptySet()
) {
    val hasMonotonicBlockBounds = (1 until blocks.size).all {
        blocks[it - 1].topPx <= blocks[it].topPx && blocks[it - 1].bottomPx <= blocks[it].bottomPx
    }
    val fragmentKinds: Set<PreparedProseFragmentKind> get() = blocks.flatMapTo(
        linkedSetOf()
    ) { block ->
        block.fragments.map { it.kind }
    }

    inline fun forEachBlockIntersecting(clip: Rect, action: (PreparedProseBlock) -> Unit) {
        if (!hasMonotonicBlockBounds) {
            for (block in blocks) if (block.intersects(clip)) action(block)
            return
        }
        var lower = 0
        var upper = blocks.size
        while (lower < upper) {
            val middle = (lower + upper) ushr 1
            if (blocks[middle].bottomPx > clip.top) upper = middle else lower = middle + 1
        }
        while (lower < blocks.size) {
            val block = blocks[lower]
            if (block.topPx >= clip.bottom) return
            action(block)
            lower += 1
        }
    }

    inline fun forEachFragmentIntersecting(clip: Rect, action: (PreparedProseFragment) -> Unit) =
        forEachBlockIntersecting(clip) { block ->
            block.fragments.filterTo(mutableListOf()) {
                Rect.intersects(it.bounds, clip)
            }.forEach(action)
        }

    companion object {
        fun error(key: ProseLayoutKey, widthPx: Int, error: ProseViewerError) =
            PreparedProseLayout(key, widthPx, 0, emptyList(), retainedBytes = 0, error = error)
    }
}

File: ios/Viewer/PreparedProseLayout.swift
import CoreText
import UIKit

/// Validates the physical-width identity before it crosses into an integer key.
enum ProseLayoutMetrics {
    static func widthPixels(widthPoints: CGFloat, scale: CGFloat) -> Int? {
        guard widthPoints.isFinite, widthPoints > 0, scale.isFinite, scale > 0 else {
            return nil
        }
        let physicalWidth = widthPoints * scale
        guard physicalWidth.isFinite, physicalWidth > 0 else { return nil }
        let roundedWidth = physicalWidth.rounded()
        // `CGFloat(Int.max)` rounds up to 2^63 on 64-bit platforms. Its predecessor
        // is the largest floating-point value that remains safe to convert to `Int`.
        guard roundedWidth.isFinite,
              roundedWidth > 0,
              roundedWidth <= CGFloat(Int.max).nextDown
        else {
            return nil
        }
        return Int(roundedWidth)
    }

    static func canonicalWidth(widthPixels: Int, scale: CGFloat) -> CGFloat {
        CGFloat(widthPixels) / scale
    }
}

struct ProseLayoutKey: Hashable {
    let semanticKey: String
    let widthPixels: Int
    let themeDigest: String
    let nativeFontRevision: UInt64
    let fontEnvironmentRevision: UInt64
    let displayScaleBits: UInt64
    let attachmentRevision: UInt64
    let generationIdentity: String
    /// Immutable semantic identity used only by generation-scoped diagnostics.
    /// It intentionally excludes layout, attachment, and font replacements.
    let semanticGenerationIdentity: String

    init(
        semanticKey: String,
        widthPixels: Int,
        themeDigest: String,
        nativeFontRevision: UInt64,
        fontEnvironmentRevision: UInt64,
        displayScale: CGFloat,
        attachmentRevision: UInt64,
        generationIdentity: String,
        semanticGenerationIdentity: String
    ) {
        self.semanticKey = semanticKey
        self.widthPixels = widthPixels
        self.themeDigest = themeDigest
        self.nativeFontRevision = nativeFontRevision
        self.fontEnvironmentRevision = fontEnvironmentRevision
        self.displayScaleBits = Double(displayScale).bitPattern
        self.attachmentRevision = attachmentRevision
        self.generationIdentity = generationIdentity
        self.semanticGenerationIdentity = semanticGenerationIdentity
    }
}

/// Fabric recycles component views, so the handoff owner must identify both
/// the mounted surface and the component instance within it.
struct FabricSurfaceToken: Hashable {
    let surfaceId: Int64
    let componentTag: Int64
}

struct FabricLeaseKey: Hashable {
    let surface: FabricSurfaceToken
    let layout: ProseLayoutKey
    /// Opaque Fabric-state incarnation handle. A recycled surface may later
    /// reuse the same component tag and generation identity; only this exact
    /// handle may consume or retire its Yoga handoff.
    let leaseHandle: UInt64
}

struct FabricGenerationToken: Hashable {
    let surface: FabricSurfaceToken
    let generationIdentity: String
    /// This is carried by custom Fabric state from the shadow-node measurement
    /// to the component view. It is never inferred from registry state.
    let leaseHandle: UInt64

    init(
        surface: FabricSurfaceToken,
        generationIdentity: String,
        leaseHandle: UInt64 = 1
    ) {
        self.surface = surface
        self.generationIdentity = generationIdentity
        self.leaseHandle = leaseHandle
    }
}

/// State-family identity across all semantic, attachment, font, and width
/// revisions. The registry permits at most one committed generation for it.
struct FabricLeaseOwner: Hashable {
    let surface: FabricSurfaceToken
    let leaseHandle: UInt64
}

/// This identity intentionally excludes the surface owner: completed immutable
/// layouts may be shared, while leases remain surface-scoped.
struct ProseMountKey: Hashable {
    let generationIdentity: String
    let widthPixels: Int
    let displayScaleBits: UInt64

    init(generationIdentity: String, widthPixels: Int, displayScale: CGFloat) {
        self.init(
            generationIdentity: generationIdentity,
            widthPixels: widthPixels,
            displayScaleBits: Double(displayScale).bitPattern
        )
    }

    init(generationIdentity: String, widthPixels: Int, displayScaleBits: UInt64) {
        self.generationIdentity = generationIdentity
        self.widthPixels = widthPixels
        self.displayScaleBits = displayScaleBits
    }
}

enum PreparedProseFragmentKind: String, Hashable {
    case text
    case marker
    case background
    case border
    case rule
    case atom
    case strike
    case image
}

/// A fully prepared paint operation. Core Text lines, colours, metrics, and
/// rectangles are all frozen before this reaches the drawing view.
final class PreparedProseFragment {
    let styleBox: EditorStyleBox?
    let kind: PreparedProseFragmentKind
    let line: CTLine?
    /// Core Text baseline measured down from the artifact's top edge.
    let origin: CGPoint
    let bounds: CGRect
    let color: CGColor?
    let borderColor: CGColor?
    let cornerRadius: CGFloat
    let strokeWidth: CGFloat
    let padding: UIEdgeInsets
    let label: String?
    let checked: Bool

    init(
        kind: PreparedProseFragmentKind,
        line: CTLine? = nil,
        origin: CGPoint = .zero,
        bounds: CGRect,
        color: CGColor? = nil,
        borderColor: CGColor? = nil,
        cornerRadius: CGFloat = 0,
        strokeWidth: CGFloat = 0,
        padding: UIEdgeInsets = .zero,
        label: String? = nil,
        checked: Bool = false,
        styleBox: EditorStyleBox? = nil
    ) {
        self.styleBox = styleBox
        self.kind = kind
        self.line = line
        self.origin = origin
        self.bounds = bounds
        self.color = color
        self.borderColor = borderColor
        self.cornerRadius = cornerRadius
        self.strokeWidth = strokeWidth
        self.padding = padding
        self.label = label
        self.checked = checked
    }

    /// Core Text retains opaque shaping data outside Swift's object graph.
    /// Charge a conservative fixed payload per line/fragment (including every
    /// prepared strike rectangle) plus the visible label so narrow documents
    /// cannot evade the prepared-layout cache.
    var estimatedRetainedBytes: Int {
        let lineBytes = line == nil ? 0 : 768
        let labelBytes = (label?.utf8.count ?? 0).layoutSaturatingMultiply(2)
        return 192 + lineBytes + labelBytes + (kind == .atom ? 192 : 0)
    }
}

/// A vertical-culling unit, sorted by its top edge. It owns every paint
/// operation needed for one semantic block so draw(_:) never shapes text.
struct PreparedProseAtomSlot {
    let nodeType: String
    let docPos: UInt32
    let attrsJSON: String
    let bounds: CGRect

    var estimatedRetainedBytes: Int { 128 + nodeType.utf8.count * 2 + attrsJSON.utf8.count * 2 }
}

final class PreparedProseBlock {
    let atomSlot: PreparedProseAtomSlot?
    let imageAttachment: ViewerImageAttachment?
    let tableBounds: CGRect?
    let fragments: [PreparedProseFragment]
    let bounds: CGRect
    let tableSurface: ViewerTableSurface?

    init(fragments: [PreparedProseFragment], bounds: CGRect, atomSlot: PreparedProseAtomSlot? = nil, imageAttachment: ViewerImageAttachment? = nil, tableSurface: ViewerTableSurface? = nil, tableBounds: CGRect? = nil) {
        self.atomSlot = atomSlot
        self.imageAttachment = imageAttachment
        self.tableBounds = tableBounds
        self.fragments = fragments
        self.bounds = bounds
        self.tableSurface = tableSurface
    }

    var estimatedRetainedBytes: Int {
        160 + fragments.reduce(0) { $0 + $1.estimatedRetainedBytes } + (atomSlot?.estimatedRetainedBytes ?? 0) + (tableSurface?.retainedBytes ?? 0)
    }

    /// Compatibility initializer retained for test seams.
    convenience init(line: CTLine, origin: CGPoint, range _: NSRange, bounds: CGRect) {
        self.init(fragments: [.init(kind: .text, line: line, origin: origin, bounds: bounds)], bounds: bounds)
    }
}

/// Immutable hit payload. Rectangles are in artifact coordinates, ordered by
/// visual line, and never require Core Text work after publication.
struct PreparedProseInteraction: Hashable {
    enum Kind: Hashable { case link, mention }

    let kind: Kind
    let rects: [CGRect]
    let href: String?
    let visibleText: String
    let docPos: UInt32?
    let label: String
    let attrsJSON: String?
    /// Local prepared-block index, retained so mounted traversal never infers source order from paint bounds.
    let sourceBlockIndex: Int?

    init(kind: Kind, rects: [CGRect], href: String?, visibleText: String, docPos: UInt32?, label: String, attrsJSON: String?, sourceBlockIndex: Int? = nil) {
        self.kind = kind
        self.rects = rects
        self.href = href
        self.visibleText = visibleText
        self.docPos = docPos
        self.label = label
        self.attrsJSON = attrsJSON
        self.sourceBlockIndex = sourceBlockIndex
    }

    var estimatedRetainedBytes: Int {
        144 + rects.count * 64 + (href?.utf8.count ?? 0) * 2 + visibleText.utf8.count * 2
            + label.utf8.count * 2 + (attrsJSON?.utf8.count ?? 0) + (sourceBlockIndex == nil ? 0 : 8)
    }
}

/// A lightweight virtual-node descriptor. UIKit elements are deliberately
/// created only when VoiceOver asks the container for an index.
struct PreparedProseAccessibilityNode: Hashable {
    enum Role: Hashable { case text, heading, link, mention, image, separator }

    let interactionIndex: Int?
    let role: Role
    let label: String
    let rects: [CGRect]
    let sourceBlockIndex: Int?

    init(interactionIndex: Int?, role: Role, label: String, bounds: CGRect, sourceBlockIndex: Int? = nil) {
        self.init(interactionIndex: interactionIndex, role: role, label: label, rects: [bounds], sourceBlockIndex: sourceBlockIndex)
    }

    init(interactionIndex: Int?, role: Role, label: String, rects: [CGRect], sourceBlockIndex: Int? = nil) {
        self.interactionIndex = interactionIndex
        self.role = role
        self.label = label
        self.rects = rects
        self.sourceBlockIndex = sourceBlockIndex
    }

    var bounds: CGRect { rects.reduce(.null) { $0.union($1) } }

    var estimatedRetainedBytes: Int { 96 + rects.count * 64 + label.utf8.count * 2 + (sourceBlockIndex == nil ? 0 : 8) }
}

public final class PreparedProseLayout: NSObject {
    let highlightingRequest: PreparedViewerHighlightingRequest?
    let highlightingResolved: Bool
    let decorations: [PreparedProseFragment]
    let key: ProseLayoutKey
    let size: CGSize
    let blocks: [PreparedProseBlock]
    let hasMonotonicBlockBounds: Bool
    let interactions: [PreparedProseInteraction]
    let accessibilityNodes: [PreparedProseAccessibilityNode]
    let imageAttachments: [ViewerImageAttachment]
    let retainedBytes: Int
    let cellShapeCatalogRetainedBytes: Int
    let error: ProseViewerError?
    let cellShape: PreparedCellShape?

    init(
        key: ProseLayoutKey,
        size: CGSize,
        blocks: [PreparedProseBlock],
        interactions: [PreparedProseInteraction] = [],
        accessibilityNodes: [PreparedProseAccessibilityNode] = [],
        imageAttachments: [ViewerImageAttachment] = [],
        retainedBytes: Int,
        error: ProseViewerError? = nil,
        decorations: [PreparedProseFragment] = [],
        highlightingRequest: PreparedViewerHighlightingRequest? = nil,
        highlightingResolved: Bool = false,
        cellShape: PreparedCellShape? = nil
    ) {
        self.highlightingRequest = highlightingRequest
        self.highlightingResolved = highlightingResolved
        self.decorations = decorations
        self.key = key
        self.size = size
        self.blocks = blocks
        self.hasMonotonicBlockBounds = zip(blocks, blocks.dropFirst()).allSatisfy {
            $0.bounds.minY <= $1.bounds.minY && $0.bounds.maxY <= $1.bounds.maxY
        }
        self.interactions = interactions
        self.accessibilityNodes = accessibilityNodes
        self.imageAttachments = imageAttachments
        self.retainedBytes = retainedBytes
        self.error = error
        self.cellShape = cellShape
        self.cellShapeCatalogRetainedBytes = PreparedProseLayout.shapeCatalogRetainedBytes(
            cellShape: cellShape,
            blocks: blocks
        )
        super.init()
    }

    func withCellShape(_ shape: PreparedCellShape) -> PreparedProseLayout {
        PreparedProseLayout(
            key: key,
            size: size,
            blocks: blocks,
            interactions: interactions,
            accessibilityNodes: accessibilityNodes,
            imageAttachments: imageAttachments,
            retainedBytes: retainedBytes,
            error: error,
            decorations: decorations,
            highlightingRequest: highlightingRequest,
            highlightingResolved: highlightingResolved,
            cellShape: shape
        )
    }

    static func error(key: ProseLayoutKey, width: CGFloat, error: ProseViewerError) -> PreparedProseLayout {
        PreparedProseLayout(key: key, size: CGSize(width: width, height: 0), blocks: [], retainedBytes: 0, error: error)
    }

    private static func shapeCatalogRetainedBytes(
        cellShape: PreparedCellShape?,
        blocks: [PreparedProseBlock]
    ) -> Int {
        var shapes: [ObjectIdentifier: PreparedCellShape] = [:]
        func collect(_ shape: PreparedCellShape?) {
            guard let shape else { return }
            shapes[ObjectIdentifier(shape)] = shape
        }
        func visit(_ layout: PreparedProseLayout) {
            collect(layout.cellShape)
            for block in layout.blocks {
                for cell in block.tableSurface?.cells ?? [] { visit(cell.content) }
            }
        }
        collect(cellShape)
        for block in blocks {
            for cell in block.tableSurface?.cells ?? [] { visit(cell.content) }
        }
        return shapes.values.reduce(0) { $0 + $1.catalogRetainedBytes }
    }
}

private extension Int {
    func layoutSaturatingMultiply(_ other: Int) -> Int {
        let result = multipliedReportingOverflow(by: other)
        return result.overflow ? Int.max : result.partialValue
    }
}
